rsocket-core/src/main/java/io/rsocket/metadata/CompositeMetadata.java

```diff
@@ -1,154 +1,215 @@
-      WellKnownMimeType wkn = WellKnownMimeType.fromId(id);
-      if (wkn != WellKnownMimeType.UNKNOWN_RESERVED_MIME_TYPE) {
-        this.mime = wkn.getMime();
-      }
-    } else {
-      this.id = -1;
-      CharSequence charSequence = CompositeMetadataFlyweight.decodeMimeTypeFromMimeBuffer(header);
-      if (charSequence == null) {
-        throw new IllegalArgumentException(
-            "composite metadata entry parsing failed on custom type");
-      }
-      this.mime = charSequence.toString();
+    @Override
+    public String getMimeType() {
+      return this.type;
     }
   }
 
   /**
-   * Returns the mime type {@link String} representation of the currently decoded entry, if there is
-   * one.
-   *
-   * <p>A {@code null} value should only occur with a positive {@link #getCurrentMimeId()}, denoting
-   * an entry that is compressed but unparseable (probably a buffer encoded by another version of
-   * the well known mime type extension spec, which is only reserved in this implementation).
-   *
-   * @return the mime type for this entry, or null
+   * An {@link Entry} backed by a {@link WellKnownMimeType} entry, but one that is not understood by
+   * this implementation.
    */
-  @Nullable
-  public String getCurrentMimeType() {
-    return this.mime;
-  }
+  public static final class ReservedMimeTypeEntry implements Entry {
+    private final ByteBuf content;
+    private final int type;
 
-  /**
-   * @return the compressed mime id byte if relevant (0-127), or -1 if not (custom mime type as
-   *     {@link String})
-   */
-  public byte getCurrentMimeId() {
-    return this.id;
+    public ReservedMimeTypeEntry(ByteBuf content, int type) {
+      this.content = content;
+      this.type = type;
+    }
+
+    @Override
+    public ByteBuf getContent() {
+      return this.content;
+    }
+
+    @Override
+    public String getMimeType() {
+      return null;
+    }
+
+    /**
+     * Returns the reserved, but unknown {@link WellKnownMimeType} for this entry.
+     *
+     * @return the reserved, but unknown {@link WellKnownMimeType} for this entry
+     */
+    public int getType() {
+      return this.type;
+    }
   }
 
-  /** @return the metadata content of the currently decoded entry */
-  public ByteBuf getCurrentContent() {
-    return this.content;
+  /** An {@link Entry} backed by a {@link WellKnownMimeType}. */
+  public static final class WellKnownMimeTypeEntry implements Entry {
+
+    private final ByteBuf content;
+    private final WellKnownMimeType type;
+
+    public WellKnownMimeTypeEntry(ByteBuf content, WellKnownMimeType type) {
+      this.content = content;
+      this.type = type;
+    }
+
+    @Override
+    public ByteBuf getContent() {
+      return this.content;
+    }
+
+    @Override
+    public String getMimeType() {
+      return this.type.getString();
+    }
+
+    /**
+     * Returns the {@link WellKnownMimeType} for this entry.
```

Comment:
Would it be worth it to state the 0-127 range? (here and in the `ReservedMimeTypeEntry`)